provider/aws/aws.go

```diff
@@ -27,41 +30,209 @@ const (
+	errRollbackComplete       = fmt.Errorf("wait for stack failed with %s", cloudformation.StackStatusRollbackComplete)
+	errUpdateRollbackComplete = fmt.Errorf("wait for stack failed with %s", cloudformation.StackStatusUpdateRollbackComplete)
+	errRollbackFailed         = fmt.Errorf("wait for stack failed with %s", cloudformation.StackStatusRollbackFailed)
+	errUpdateRollbackFailed   = fmt.Errorf("wait for stack failed with %s", cloudformation.StackStatusUpdateRollbackFailed)
+	errDeleteFailed           = fmt.Errorf("wait for stack failed with %s", cloudformation.StackStatusDeleteFailed)
+	errTimeoutExceeded        = fmt.Errorf("wait for stack timeout exceeded")
+)
+
+type AWSProvider struct {
 	dry                        bool
 	vpcID                      string
 	natCidrBlocks              []string
 	availabilityZones          []string
 	cloudformation             cloudformationiface.CloudFormationAPI
 	ec2                        ec2iface.EC2API
 	stackTerminationProtection bool
+	logger                     *log.Entry
 }
 
-func NewAwsProvider(dry bool, vpcID string, natCidrBlocks, availabilityZones []string, stackTerminationProtection bool) *AwsProvider {
+func NewAWSProvider(dry bool, vpcID string, natCidrBlocks, availabilityZones []string, stackTerminationProtection bool) *AWSProvider {
+	// TODO: find vpcID at startup
 	p := defaultConfigProvider()
-	return &AwsProvider{
+	return &AWSProvider{
 		dry:                        dry,
 		vpcID:                      vpcID,
 		natCidrBlocks:              natCidrBlocks,
 		availabilityZones:          availabilityZones,
 		cloudformation:             cloudformation.New(p),
 		ec2:                        ec2.New(p),
 		stackTerminationProtection: stackTerminationProtection,
+		logger:                     log.WithFields(log.Fields{"provider": ProviderName}),
 	}
 }
 
-func (p AwsProvider) String() string {
+func (p AWSProvider) String() string {
 	return ProviderName
 }
 
-func (p *AwsProvider) generateStackSpec(nets []string) (*stackSpec, error) {
+func (p *AWSProvider) Ensure(configs map[provider.Resource]map[string]struct{}) error {
+	params := &cloudformation.DescribeStacksInput{
+		StackName: aws.String(stackName),
+	}
+	resp, err := p.cloudformation.DescribeStacks(params)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); !ok || aerr.Code() == cloudformation.ErrCodeStackInstanceNotFoundException {
+			return err
+		}
+	}
+
+	// don't do anything if the stack doesn't exist and the config is empty
+	if len(configs) == 0 && len(resp.Stacks) == 0 {
+		return nil
+	}
+
+	spec, err := p.generateStackSpec(configs)
+	if err != nil {
+		return errors.Wrap(err, "failed to generate stack spec")
+	}
+
+	// create new stack if it doesn't already exists
+	if len(resp.Stacks) == 0 {
+		p.logger.Infof("Creating CF stack with config: %v", configs)
+		err := p.createCFStack(spec)
+		if err != nil {
+			return errors.Wrap(err, "failed to create CF stack")
```

Comment:
I guess this will be triggered on every resync. We might want to test it in a cluster and check if we can find out the error that we want to silently drop.